gossip/dagordering/event_buffer.go

```diff
@@ -0,0 +1,198 @@
+	}
+
+	buf.mu.Lock()
+	defer buf.mu.Unlock()
+
+	if _, ok := buf.incompletes.Peek(e.event.ID()); ok {
+		// duplicate
+		buf.dropEvent(e, eventcheck.ErrDuplicateEvent)
+		buf.releaseEvent(e)
+		return false
+	}
+	complete = buf.pushEvent(e, nil, false)
+	buf.spillIncompletes(buf.limit)
+	return complete
+}
+
+func (buf *EventsBuffer) pushEvent(e *event, incompleteEventsList []*event, recheck bool) bool {
+	if buf.callback.Exists(e.event.ID()) {
+		buf.incompletes.Remove(e.event.ID())
+		if !recheck {
+			buf.dropEvent(e, eventcheck.ErrAlreadyConnectedEvent)
+		}
+		buf.releaseEvent(e)
+		return false
+	}
+	parents := buf.completeEventParents(e)
+	if parents == nil {
+		if !recheck {
+			buf.incompletes.Add(e.event.ID(), e, uint(e.event.Size()))
+		}
+		return false
+	}
+
+	ok := buf.processCompleteEvent(e, parents)
+	buf.releaseEvent(e)
+
+	if ok {
+		// now child events may become complete, check it again
+		eHash := e.event.ID()
+		if incompleteEventsList == nil {
+			incompleteEventsList = buf.getIncompleteEventsList()
+		}
+		for _, child := range incompleteEventsList {
+			for _, parent := range child.event.Parents() {
+				if parent == eHash {
+					buf.pushEvent(child, incompleteEventsList, true)
+					break
+				}
+			}
+		}
+	}
+	buf.incompletes.Remove(e.event.ID())
+	return ok
+}
+
+func (buf *EventsBuffer) getIncompleteEventsList() []*event {
+	res := make([]*event, 0, buf.incompletes.Len())
+	for _, childID := range buf.incompletes.Keys() {
+		child, _ := buf.incompletes.Peek(childID)
+		if child == nil {
+			continue
+		}
+		res = append(res, child.(*event))
+	}
+	return res
+}
+
+func (buf *EventsBuffer) completeEventParents(e *event) dag.Events {
+	parents := make(dag.Events, len(e.event.Parents()))
+	for i, p := range e.event.Parents() {
+		parent := buf.callback.Get(p)
+		if parent == nil {
+			return nil
+		}
+		parents[i] = parent
+	}
+	return parents
+}
+
+func (buf *EventsBuffer) processCompleteEvent(e *event, parents dag.Events) bool {
+	// validate
+	if buf.callback.Check != nil {
+		err := buf.callback.Check(e.event, parents)
+		if err != nil {
+			buf.dropEvent(e, err)
+			return false
+		}
+	}
+
+	// process
+	err := buf.callback.Process(e.event)
+	if err != nil {
+		e.err = err
+		buf.dropEvent(e, err)
+		return false
+	}
+	return true
+}
+
+func (buf *EventsBuffer) spillIncompletes(limit dag.Metric) {
+	for idx.Event(buf.incompletes.Len()) > limit.Num || uint64(buf.incompletes.Weight()) > limit.Size {
+		_, val, ok := buf.incompletes.RemoveOldest()
+		if !ok {
+			break
+		}
+		e := val.(*event)
+		buf.dropEvent(e, eventcheck.ErrSpilledEvent)
+		buf.releaseEvent(e)
+	}
+}
+
+func (buf *EventsBuffer) dropEvent(e *event, err error) {
+	if e.err == nil {
+		e.err = err
+	}
+}
+
+func (buf *EventsBuffer) releaseEvent(e *event) {
+	if buf.callback.Released != nil && !e.released {
+		buf.callback.Released(e.event, e.peer, e.err)
+	}
+	e.released = true
+}
+
+func (buf *EventsBuffer) IsBuffered(id hash.Event) bool {
```

Comment:
```suggestion
func (buf *EventsBuffer) Contains(id hash.Event) bool {
```

(is buffered can be confused with configuration of IO or channels, where being buffered means something else)